Panic in the `PbGenericBlock()`

### Describe the bug

If I substitute `params.BeaconConfig().GenesisForkVersion` with `params.BeaconConfig().DenebForkVersion` in this test
```
tt.action()
bFunc, ok := BlockMap[bytesutil.ToBytes4(params.BeaconConfig().GenesisForkVersion)]
assert.Equal(t, tt.exists, ok)
if tt.exists {
	b, err := bFunc()
	require.NoError(t, err)
	generic, err := b.PbGenericBlock()
	require.NoError(t, err)
	assert.NotNil(t, generic.GetPhase0())
}
```
the test will panic [on line 128](https://github.com/prysmaticlabs/prysm/blob/c48d40907c2652c99705a8148542d0904e135ac8/consensus-types/blocks/getters.go#L128)

### Has this worked before in a previous version?

```text
Yes, this worked in `v5.0.3`
```


### 🔬 Minimal Reproduction

_No response_

### Error

```text
=== RUN   TestInitializeDataMaps
=== RUN   TestInitializeDataMaps/no_change
--- FAIL: TestInitializeDataMaps/no_change (18.89s)

--- FAIL: TestInitializeDataMaps (18.89s)
panic: interface conversion: protoreflect.ProtoMessage is *eth.SignedBeaconBlockDeneb, not *eth.SignedBeaconBlockContentsDeneb [recovered]
	panic: interface conversion: protoreflect.ProtoMessage is *eth.SignedBeaconBlockDeneb, not *eth.SignedBeaconBlockContentsDeneb

goroutine 36 [running]:
testing.tRunner.func1.2({0x100b3fd20, 0x14000418fc0})
	/Users/bogdan/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.4.darwin-arm64/src/testing/testing.go:1632 +0x3c0
testing.tRunner.func1()
	/Users/bogdan/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.4.darwin-arm64/src/testing/testing.go:1635 +0x57c
panic({0x100b3fd20?, 0x14000418fc0?})
	/Users/bogdan/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.4.darwin-arm64/src/runtime/panic.go:785 +0xf0
github.com/prysmaticlabs/prysm/v5/consensus-types/blocks.(*SignedBeaconBlock).PbGenericBlock(0x140002be5b0)
	/Users/bogdan/go/src/github.com/prysmaticlabs/prysm/consensus-types/blocks/getters.go:128 +0x1014
github.com/prysmaticlabs/prysm/v5/beacon-chain/p2p/types.TestInitializeDataMaps.func4(0x140002b6d00)
	/Users/bogdan/go/src/github.com/prysmaticlabs/prysm/beacon-chain/p2p/types/object_mapping_test.go:57 +0x228
testing.tRunner(0x140002b6d00, 0x140003ba1e0)
	/Users/bogdan/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.4.darwin-arm64/src/testing/testing.go:1690 +0x1a8
created by testing.(*T).Run in goroutine 35
	/Users/bogdan/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.4.darwin-arm64/src/testing/testing.go:1743 +0x668

Debugger finished with the exit code 0
```
```


### Platform(s)

Mac (Apple Silicon)

### What version of Prysm are you running? (Which release)

v5.2.0

### Anything else relevant (validator index / public key)?

_No response_

> ### Describe the bug
> If I substitute `params.BeaconConfig().GenesisForkVersion` with `params.BeaconConfig().DenebForkVersion` in this test
> 
> ```
> tt.action()
> bFunc, ok := BlockMap[bytesutil.ToBytes4(params.BeaconConfig().GenesisForkVersion)]
> assert.Equal(t, tt.exists, ok)
> if tt.exists {
> 	b, err := bFunc()
> 	require.NoError(t, err)
> 	generic, err := b.PbGenericBlock()
> 	require.NoError(t, err)
> 	assert.NotNil(t, generic.GetPhase0())
> }
> ```
> 
> the test will panic [on line 128](https://github.com/prysmaticlabs/prysm/blob/c48d40907c2652c99705a8148542d0904e135ac8/consensus-types/blocks/getters.go#L128)
> 
> ### Has this worked before in a previous version?
> ```
> Yes, this worked in `v5.0.3`
> ```
> 
> ### 🔬 Minimal Reproduction
> _No response_
> 
> ### Error
> ```
> === RUN   TestInitializeDataMaps
> === RUN   TestInitializeDataMaps/no_change
> --- FAIL: TestInitializeDataMaps/no_change (18.89s)
> 
> --- FAIL: TestInitializeDataMaps (18.89s)
> panic: interface conversion: protoreflect.ProtoMessage is *eth.SignedBeaconBlockDeneb, not *eth.SignedBeaconBlockContentsDeneb [recovered]
> 	panic: interface conversion: protoreflect.ProtoMessage is *eth.SignedBeaconBlockDeneb, not *eth.SignedBeaconBlockContentsDeneb
> 
> goroutine 36 [running]:
> testing.tRunner.func1.2({0x100b3fd20, 0x14000418fc0})
> 	/Users/bogdan/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.4.darwin-arm64/src/testing/testing.go:1632 +0x3c0
> testing.tRunner.func1()
> 	/Users/bogdan/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.4.darwin-arm64/src/testing/testing.go:1635 +0x57c
> panic({0x100b3fd20?, 0x14000418fc0?})
> 	/Users/bogdan/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.4.darwin-arm64/src/runtime/panic.go:785 +0xf0
> github.com/prysmaticlabs/prysm/v5/consensus-types/blocks.(*SignedBeaconBlock).PbGenericBlock(0x140002be5b0)
> 	/Users/bogdan/go/src/github.com/prysmaticlabs/prysm/consensus-types/blocks/getters.go:128 +0x1014
> github.com/prysmaticlabs/prysm/v5/beacon-chain/p2p/types.TestInitializeDataMaps.func4(0x140002b6d00)
> 	/Users/bogdan/go/src/github.com/prysmaticlabs/prysm/beacon-chain/p2p/types/object_mapping_test.go:57 +0x228
> testing.tRunner(0x140002b6d00, 0x140003ba1e0)
> 	/Users/bogdan/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.4.darwin-arm64/src/testing/testing.go:1690 +0x1a8
> created by testing.(*T).Run in goroutine 35
> 	/Users/bogdan/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.4.darwin-arm64/src/testing/testing.go:1743 +0x668
> 
> Debugger finished with the exit code 0
> ```
> 
> ```
> 
> 
> ### Platform(s)
> 
> Mac (Apple Silicon)
> 
> ### What version of Prysm are you running? (Which release)
> 
> v5.2.0
> 
> ### Anything else relevant (validator index / public key)?
> 
> _No response_
> ```

--- a/changelog/james-prysm_handle-pbgeneric-error.md
+++ b/changelog/james-prysm_handle-pbgeneric-error.md
@@ -1,0 +1,3 @@
+### Fixed
+
+- fix panic with type cast on pbgenericblock()

--- a/consensus-types/blocks/getters.go
+++ b/consensus-types/blocks/getters.go
@@ -124,8 +124,12 @@ func (b *SignedBeaconBlock) PbGenericBlock() (*eth.GenericSignedBeaconBlock, err
 				Block: &eth.GenericSignedBeaconBlock_BlindedDeneb{BlindedDeneb: pb.(*eth.SignedBlindedBeaconBlockDeneb)},
 			}, nil
 		}
+		bc, ok := pb.(*eth.SignedBeaconBlockContentsDeneb)
+		if !ok {
+			return nil, fmt.Errorf("PbGenericBlock() only supports block content type but got %T", pb)
+		}
 		return &eth.GenericSignedBeaconBlock{
-			Block: &eth.GenericSignedBeaconBlock_Deneb{Deneb: pb.(*eth.SignedBeaconBlockContentsDeneb)},
+			Block: &eth.GenericSignedBeaconBlock_Deneb{Deneb: bc},
 		}, nil
 	case version.Electra:
 		if b.IsBlinded() {
@@ -133,8 +137,12 @@ func (b *SignedBeaconBlock) PbGenericBlock() (*eth.GenericSignedBeaconBlock, err
 				Block: &eth.GenericSignedBeaconBlock_BlindedElectra{BlindedElectra: pb.(*eth.SignedBlindedBeaconBlockElectra)},
 			}, nil
 		}
+		bc, ok := pb.(*eth.SignedBeaconBlockContentsElectra)
+		if !ok {
+			return nil, fmt.Errorf("PbGenericBlock() only supports block content type but got %T", pb)
+		}
 		return &eth.GenericSignedBeaconBlock{
-			Block: &eth.GenericSignedBeaconBlock_Electra{Electra: pb.(*eth.SignedBeaconBlockContentsElectra)},
+			Block: &eth.GenericSignedBeaconBlock_Electra{Electra: bc},
 		}, nil
 	case version.Fulu:
 		if b.IsBlinded() {
@@ -142,8 +150,12 @@ func (b *SignedBeaconBlock) PbGenericBlock() (*eth.GenericSignedBeaconBlock, err
 				Block: &eth.GenericSignedBeaconBlock_BlindedFulu{BlindedFulu: pb.(*eth.SignedBlindedBeaconBlockFulu)},
 			}, nil
 		}
+		bc, ok := pb.(*eth.SignedBeaconBlockContentsFulu)
+		if !ok {
+			return nil, fmt.Errorf("PbGenericBlock() only supports block content type but got %T", pb)
+		}
 		return &eth.GenericSignedBeaconBlock{
-			Block: &eth.GenericSignedBeaconBlock_Fulu{Fulu: pb.(*eth.SignedBeaconBlockContentsFulu)},
+			Block: &eth.GenericSignedBeaconBlock_Fulu{Fulu: bc},
 		}, nil
 	default:
 		return nil, errIncorrectBlockVersion

--- a/testing/util/block_test.go
+++ b/testing/util/block_test.go
@@ -322,3 +322,30 @@ func TestGenerateVoluntaryExits(t *testing.T) {
 	require.NoError(t, err)
 	require.NoError(t, coreBlock.VerifyExitAndSignature(val, beaconState, exit))
 }
+
+func Test_PostDenebPbGenericBlock_ErrorsForPlainBlock(t *testing.T) {
+	t.Run("Deneb block returns type error", func(t *testing.T) {
+		eb := NewBeaconBlockDeneb()
+		b, err := blocks.NewSignedBeaconBlock(eb)
+		require.NoError(t, err)
+
+		_, err = b.PbGenericBlock()
+		require.ErrorContains(t, "PbGenericBlock() only supports block content type but got", err)
+	})
+	t.Run("Electra block returns type error", func(t *testing.T) {
+		eb := NewBeaconBlockElectra()
+		b, err := blocks.NewSignedBeaconBlock(eb)
+		require.NoError(t, err)
+
+		_, err = b.PbGenericBlock()
+		require.ErrorContains(t, "PbGenericBlock() only supports block content type but got", err)
+	})
+	t.Run("Fulu block returns type error", func(t *testing.T) {
+		eb := NewBeaconBlockFulu()
+		b, err := blocks.NewSignedBeaconBlock(eb)
+		require.NoError(t, err)
+
+		_, err = b.PbGenericBlock()
+		require.ErrorContains(t, "PbGenericBlock() only supports block content type but got", err)
+	})
+}

